pipelines/whale/extractor/spanner_metadata_extractor.py

```diff
@@ -0,0 +1,152 @@
+        {
+            CONNECTION_NAME_KEY: "spanner",
+            WHERE_CLAUSE_SUFFIX_KEY: "",
+        }
+    )
+
+    def init(self, conf):
+        self.conf = conf.with_fallback(SpannerMetadataExtractor.DEFAULT_CONFIG)
+        self._project_id = conf.get_string(SpannerMetadataExtractor.PROJECT_ID_KEY)
+        self._connection_name = conf.get_string(
+            SpannerMetadataExtractor.CONNECTION_NAME_KEY
+        )
+        self._instance_id = self.conf.get_string(
+            SpannerMetadataExtractor.INSTANCE_ID_KEY
+        )
+        self._database_id = self.conf.get(
+            SpannerMetadataExtractor.DATABASE_ID_KEY, None
+        )
+        self._key_path = self.conf.get(SpannerMetadataExtractor.KEY_PATH_KEY, None)
+
+        client_kwargs = {"project": self._project_id}
+        if self._key_path is not None:
+            spanner_client = spanner.Client.from_service_account_json(
+                self._key_path, **client_kwargs
+            )
+        else:
+            spanner_client = spanner.Client(**client_kwargs)
+
+        self.instance = spanner_client.instance(self._instance_id)
+        self.database = self.instance.database(self._database_id)
+
+        self.sql_stmt = SpannerMetadataExtractor.SQL_STATEMENT.format(
+            where_clause_suffix=self.conf.get_string(
+                SpannerMetadataExtractor.WHERE_CLAUSE_SUFFIX_KEY
+            )
+        )
+
+        LOGGER.info("SQL for Spanner: {}".format(self.sql_stmt))
+
+        self._extract_iter = None  # type: Union[None, Iterator]
+
+    def extract(self):
+        # type: () -> Union[TableMetadata, None]
+        if not self._extract_iter:
+            self._extract_iter = self._get_extract_iter()
+        try:
+            return next(self._extract_iter)
+        except StopIteration:
+            return None
+
+    def get_scope(self):
+        # type: () -> str
+        return "extractor.spanner_table_metadata"
+
+    def _get_extract_iter(self):
+        # type: () -> Iterator[TableMetadata]
+        """
+        Using itertools.groupby and raw level iterator, it groups to table and yields TableMetadata
+        :return:
+        """
+
+        with self.database.snapshot() as snapshot:
+            results = snapshot.execute_sql(self.sql_stmt)
+            header = SpannerMetadataExtractor.HEADER
+            headered_results = [dict(zip(header, result)) for result in results]
+
+            for _, group in groupby(headered_results, self._get_table_key):
+                columns = []
+
+                for row in group:
+                    last_row = row
+                    columns.append(
+                        ColumnMetadata(
+                            row["col_name"],
+                            None,
+                            row["col_type"],
+                            row["col_sort_order"],
+                        )
+                    )
+
+                schema = "{}.{}".format(self._instance_id, self._database_id)
```

Comment:
We could move this to line 118, doesn't need to be within the for loops 😊

Reply:
Good call.